### PR TITLE
ci: run nix-managed lints on PRs

### PR DESCRIPTION
**Pull request by ysndr**
Adds an action that runs "pre-commit" lints on the current PR.
pre-pommit only runs tests if relevant files.

This came up in a [comment](https://github.com/flox/flox/pull/1371#discussion_r1580874973) on #1371.

### CODE REVIEW

File: .github/workflows/nix-managed-lints.yml [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+        run: |
+          git fetch origin '${{ github.base_ref }}';
+
+      - name: "Run Nix Git Hooks"
+        run: |
+          nix develop -L --no-update-lock-file --command \
+            pre-commit run \
+            --hook-stage manual \
+            --from-ref "$( git rev-parse origin/${{ github.base_ref }} )" \
+            --to-ref   "$( git rev-parse ${{ github.head_ref }} )";
```
Comment (dcarley) [inline .github/workflows/nix-managed-lints.yml]: RE the build failures, checking all of the files is pretty fast:
```
% pre-commit run --hook-stage manual --all-files
alejandra................................................................Passed
clang-format.............................................................Passed
clippy...................................................................Passed
rustfmt..................................................................Passed

~/projects/flox/flox on HEAD (790a35d2) [$!] took 2s
```

So we could simplify it to:
```suggestion
      - name: "Run Nix Git Hooks"
        run: |
          nix develop -L --no-update-lock-file --command \
            pre-commit run \
            --hook-stage manual \
            --all-files
```

File: .github/workflows/nix-managed-lints.yml [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+      - name: "Setup Nix"
+        uses: "./.github/actions/common-setup"
+        with:
+          GITHUB_ACCESS_TOKEN: "${{ secrets.NIX_GIT_TOKEN }}"
+          SUBSTITUTER: "${{ vars.FLOX_CACHE_PUBLIC_BUCKET }}"
+          SUBSTITUTER_KEY: "${{ secrets.FLOX_CACHE_PUBLIC_NIX_SECRET_KEY }}"
+          AWS_ACCESS_KEY_ID: "${{ secrets.FLOX_CACHE_PUBLIC_AWS_ACCESS_KEY_ID }}"
+          AWS_SECRET_ACCESS_KEY: "${{ secrets.FLOX_CACHE_PUBLIC_AWS_SECRET_ACCESS_KEY }}"
+          SSH_KEY: "${{ secrets.FLOXBOT_SSH_KEY }}"
```
Comment (dcarley) [inline .github/workflows/nix-managed-lints.yml:32]: (curious, non-blocking) What are these needed for?

Comment (ysndr) [reply .github/workflows/nix-managed-lints.yml:32]: Access to the nix binary cache that we host on our s3
I dont think they are needed here, but i also didnt check whether they are truly optional in the common-setup action

File: .github/workflows/nix-managed-lints.yml (hunk 1 above)
Comment (ysndr) [reply .github/workflows/nix-managed-lints.yml]: its fast for you bc you probably have a warm rust build cache
clippy will take a while if it needs to build that cache, so skipping clippy if rust files didnt change speeds up the lint

Comment (ysndr) [reply .github/workflows/nix-managed-lints.yml]: the build failures were weird. I think its due to how the checkout action checks out the repo but still puzzling.

Comment (dcarley) [reply .github/workflows/nix-managed-lints.yml]: Ahh. Should this workflow be using the cache from here, or maybe just restore-only? https://github.com/flox/flox/blob/790a35d29d62e751f2808039e025c750f13c5502/.github/workflows/ci.yml#L138-L148

Comment (ysndr) [reply .github/workflows/nix-managed-lints.yml]: yeah possibly that would probably speed things up as well
(to a point where we may get away with `--all-files` yes).
I didnt want to spend too much time optimizing this action,
given that it's just broken out of a conversation elsewhere.
If you feel we should do more in that regard,
lets at least open an issue for that.

Comment (dcarley) [reply .github/workflows/nix-managed-lints.yml]: The git commands are a bit of a mouthful but I'm fine to merge as-is now that it's working 👍🏻

**Review by dcarley — APPROVED** [review]
Thanks for breaking this out!

**Review by mkenigs — APPROVED** [review]
I agree it may make sense in the future to look at caching and using `--all-files`. If we do that would be nice to share with other CI jobs as well

**Pull request merged**